MINOR : KIP-297 : use service loader for ConfigProvider impls

rayokota opened the pull request:
This is a small change to use the Java ServiceLoader to load ConfigProvider plugins.  It uses code added by @mageshn for Connect Rest Extensions.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

rayokota committed:
MINOR: use service loader for ConfigProvider impls

mageshn commented on clients/src/main/java/org/apache/kafka/common/config/FileConfigProvider.java:
```diff
@@ -16,6 +16,7 @@
  */
 package org.apache.kafka.common.config;
 
+import java.io.BufferedReader;
```
Should FileConfigProvider class be listed as whitelist in PluginUtils?

mageshn replied:
Also, maybe we need a Manifest for FileConfigProvider

mageshn commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java:
```diff
@@ -192,16 +192,20 @@
-            + "Typically used to add custom capability like logging, security, etc.";
+            + "<code>ConnectRestExtension</code> allows you to inject into Connect's REST API user defined resources like filters. "
+            + "Typically used to add custom capability like logging, security, etc. "
+            + "The implementations will be discovered using the standard Java <code>ServiceLoader</code> mechanism.";
```
Like this 👍

rhauch replied:
I suggest removing this last sentence, too, for the same reason.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java:
```diff
@@ -192,16 +192,20 @@
+            + "in the order specified. Implementing the interface  "
+            + "<code>ConfigProvider</code> allows you to replace variable references in connector configurations, "
+            + "such as for externalized secrets. "
+            + "The implementations will be discovered using the standard Java <code>ServiceLoader</code> mechanism.";
```
This last sentence is probably not for the user audience that will simply try to configure a worker. I suggest removing it.

rhauch reviewed (commented):
Two minor suggestions. Otherwise looks good.

mageshn reviewed (commented):
Thanks, @rayokota. Just one minor comment. Also, have you considered using <T> T newPlugin(String klassName, AbstractConfig config, Class<T> pluginKlass) in lieu of newConfigProvider? Maybe the only change that is needed is if you would want to be able to configure just configs with a prefix.

rayokota committed:
Minor doc change

rayokota committed:
Add FileConfigProvider to manifest of connect-runtime

rayokota committed:
Move the manifest for FileConfigProvider to client jar

rayokota commented:
@rhauch , I removed the comments as you suggested.  @mageshn , I added a manifest to the Kafka client jar.  The `newConfigProvider` does a bit more than `newPlugin` so I left it for now.

mageshn commented:
Thanks, @rayokota how about listing  the FileConfigProvider class as whitelist in PluginUtils

rayokota committed:
Add org.apache.kafka.common.config..*ConfigProvider to plugin WHITE_LIST

rayokota commented:
@mageshn , I added a pattern for ConfigProviders to the WHITE_LIST

mageshn reviewed (approved):
LGTM. Thanks, @rayokota.

rayokota commented:
Hi @ewencp , if you get a chance, do you think you can take a look and merge if this looks ok?

rayokota committed:
Add license to manifest file

rhauch reviewed (approved):
LGTM.

ewencp commented:
LGTM, merging to trunk and 2.0

rayokota commented:
Thanks @ewencp !
